refactor(bindings/C): Alter naming convention for consistency

Ji-Xinyou opened the pull request:
We remove `blocking` for consistency and will add async support in the future, where we will add async keyword to the API.

This PR also rename the `opendal_reader_read()` first argument from `self` to `reader`, since `self` is not a C convention.

Ji-Xinyou commented:
Relate #3281

Ji-Xinyou commented:
Oh, I notice this probably could break the swift, go and zig as well, @Xuanwo . Would you please take a look at that when you got time?

jiaoew1991 commented on bindings/c/include/opendal.h:
```diff
@@ -405,17 +405,17 @@ typedef struct opendal_result_stat {
- * For examples, please see the comment section of opendal_operator_blocking_list()
- * @see opendal_operator_blocking_list()
+ * For examples, please see the comment section of opendal_operator_list()
+ * @see opendal_operator_list()
  */
 typedef struct opendal_blocking_lister {
   struct BlockingLister *inner;
 } opendal_blocking_lister;
 
 /**
```
should it rename to `opendal_lister` ?

Ji-Xinyou replied:
Yes, nice catch on this

jiaoew1991 commented:
> Oh, I notice this probably could break the swift, go and zig as well, @Xuanwo . Would you please take a look at that when you got time?

BTW c binding is used by other language bindings, does it ready to release? 🤔

Ji-Xinyou commented:
> BTW c binding is used by other language bindings, does it ready to release? 🤔

Hmm I think the major functionalities are already there, `list`, `read`, `write`. But there are still many other functionalities that we do not support.

There are a couple of things to consider
1. Does C binding always stay sync with other bindings? => For this question I think the answer is no. I would say other bindings that depend on C binding release with a certain version of C binding.
2. Do we need to align all functionalities that are implemented by other **released** bindings?

If the answer to the second question is yes, we may need to start figuring out what we are missing.
If the answer is no, i.e. we implement them in the future release, I think the C binding is ready for a release.

In any way this needs some discussions, but it is good to mention this issue up since we do need to make a release soon 🤣

Xuanwo commented:
> Does C binding always stay sync with other bindings?

For now, YES. For always, I don't know either. It depends on the future development of OpenDAL.

Pin exact tags of C binding for other binding is complex and makes the development work much harder.

Xuanwo commented on bindings/c/include/opendal.h:
```diff
@@ -539,9 +539,9 @@ struct opendal_result_operator_new opendal_operator_new(const char *scheme,
-struct opendal_error *opendal_operator_blocking_write(const struct opendal_operator_ptr *ptr,
-                                                      const char *path,
-                                                      struct opendal_bytes bytes);
+struct opendal_error *opendal_operator_write(const struct opendal_operator_ptr *ptr,
```
How about renaming `opendal_operator_ptr` to `opendal_operator` instead? We design this name when only operator is a pointer, but now, all struct is a poiner instead. I believe that this renaming make our API more consistent.

And we can call the `opendal_operator` as `op` for short.

Xuanwo commented on bindings/c/include/opendal.h:
```diff
@@ -612,7 +612,7 @@ struct opendal_result_read opendal_operator_blocking_read(const struct opendal_o
  * int length = 13;
  * unsigned char buffer[length];
- * opendal_code r = opendal_operator_blocking_read_with_buffer(ptr, "testpath", buffer, length);
+ * opendal_code r = opendal_operator_read_with_buffer(ptr, "testpath", buffer, length);
```
typo?

Xuanwo commented on bindings/c/include/opendal.h:
```diff
@@ -628,8 +628,8 @@ struct opendal_result_read opendal_operator_blocking_read(const struct opendal_o
  */
-struct opendal_result_reader opendal_operator_blocking_reader(const struct opendal_operator_ptr *ptr,
-                                                              const char *path);
+struct opendal_result_reader opendal_operator_reader(const struct opendal_operator_ptr *ptr,
```
I'm not sure do we need to naming it as `opendal_result_operator_reader`? To match the same style like `opendal_result_reader_read`.

Ji-Xinyou replied:
I agree with you

Xuanwo commented on bindings/c/include/opendal.h:
```diff
@@ -940,15 +940,15 @@ void opendal_operator_options_free(const struct opendal_operator_options *option
+ * @see opendal_operator_list()
  */
-struct opendal_list_entry *opendal_lister_next(const struct opendal_blocking_lister *self);
+struct opendal_list_entry *opendal_lister_next(const struct opendal_lister *self);
```
lister_next() could return an error too.

Xuanwo commented on bindings/c/include/opendal.h:
```diff
@@ -809,8 +809,8 @@ struct opendal_result_stat opendal_operator_stat(const struct opendal_operator_p
  */
-struct opendal_result_list opendal_operator_blocking_list(const struct opendal_operator_ptr *ptr,
-                                                          const char *path);
+struct opendal_result_list opendal_operator_list(const struct opendal_operator_ptr *ptr,
```
Please note that our core API has been changed:

- list() returns a list of entries.
- lister() returns a lister.

So this API should be renamed to `opendal_operator_lister` and should add a new API called `opendal_operator_list`.

Ji-Xinyou replied:
I will address this later on

Xuanwo commented on bindings/c/include/opendal.h:
```diff
@@ -975,7 +975,7 @@ char *opendal_list_entry_name(const struct opendal_list_entry *self);
  */
 void opendal_list_entry_free(struct opendal_list_entry *ptr);
```
OpenDAL doesn't have a struct called `list_entry`. How about naming it as `opendal_entry`?

Ji-Xinyou replied:
I think `opendal_list_entry` is more clear since it notes that this entry comes from a list operation, otherwise it would might be a little bit confusing that the struct name does not indicate where the entries is from.

But for the purpose of aligning naming with the Rust core, `opendal_entry` makes more sense.

WDYT?

Xuanwo replied:
I prefer using `opendal_entry`.

Xuanwo reviewed (commented):
I'm willing to help address the issues for zig/go/swift. However, I would prefer to address the following comments first:

Xuanwo commented:
This is a breaking change, it's better to record it in C binding's `upgrade.md`.

Ji-Xinyou commented:
> This is a breaking change, it's better to record it in C binding's `upgrade.md`.
I will address these issues later today. Thanks for the review.

jiaoew1991 commented:
Looking forward to the API stabilizing, so that I can continue with the integration work of Milvus. 😄

Xuanwo commented on bindings/c/README.md:
```diff
@@ -14,8 +14,9 @@ A simple read and write example
-    opendal_operator_ptr op = opendal_operator_new("memory", 0);
-    assert(op.ptr != NULL);
+    opendal_result_operator_new result = opendal_operator_new("memory", 0);
+    assert(result.operator_ptr != NULL);
```
Maybe this should be `result.operator != NULL`?

Xuanwo commented:
> Looking forward to the API stabilizing, so that I can continue with the integration work of Milvus. 😄

I believe this PR will address most of the API flavor issues.

Ji-Xinyou commented:
This seems to be a PR that is related to many things, I will mark this as draft and make this ready for review after some polishments.

Xuanwo commented:
cc @Ji-Xinyou, would you like to take a review? Thanks!

Ji-Xinyou commented:
> cc @jiaoew1991, I have added an `upgrade.md` docs for C binding, would you like to take a look?
> 
> https://github.com/apache/incubator-opendal/blob/8396c4398af9281cbcbc4258e8773b05e85e75cf/bindings/c/upgrade.md

Added commits look perfect to me.

Xuanwo commented:
cc @Ji-Xinyou, do you think this PR is good to go?

Ji-Xinyou commented:
> cc @Ji-Xinyou, do you think this PR is good to go?

Yes, I think this PR is ready :)
